### ISSUE TITLE
Update CMake build to define RUNPREFIX different than CMAKE_INSTALL_PREFIX for package building compatibility

### ISSUE THREAD

**Issue by amessina** [MythTV/mythtv#980]
**Is your feature request related to a problem? Please describe.**
The CMake build assumes write access to `CMAKE_INSTALL_PREFIX`.  Packages are typically built for redistribution and prepend `DESTDIR` to `CMAKE_INSTALL_PREFIX`.  In that way, the package can be "installed" in a temporary buildroot while maintaining the runtime prefix of `CMAKE_INSTALL_PREFIX` after they are packaged and installed on other systems.

**Describe the solution you'd like**
The CMake system needs to be updated to install to `${DESTDIR}${CMAKE_INSTALL_PREFIX}` during the build so it can be packaged properly.

**Describe alternatives you've considered**
I've tried setting `CMAKE_INSTALL_PREFIX` to a place in the builddir (`/builddir/build/BUILD/mythtv-34.0-build/install`) then moving that install location to `/usr` which fails at runtime (on another computer) as the runtime prefix is also set to `/builddir/build/BUILD/mythtv-34.0-build/install` and I get errors like the following as of course it should be looking in `/usr/lib64/mythtv/plugins`:

```
mythfrontend[1320]: No plugins directory /builddir/build/BUILD/mythtv-34.0-build/install/lib64/mythtv/plugins
```

**Comment by linuxdude42** [comment]
How about this?  Does this give you the control you need?

```diff --git a/cmake/MythOptions.cmake b/cmake/MythOptions.cmake
index 1dfac3b601..ebf22af5ee 100644
--- a/cmake/MythOptions.cmake
+++ b/cmake/MythOptions.cmake
@@ -21,6 +21,11 @@ set(MYTH_DEFAULT_LIBS_PREFIX
     CACHE PATH
           "Default library install prefix if not specified on the command line."
 )
+set(MYTH_RUN_PREFIX
+    ""
+    CACHE PATH
+          "The prefix where MythTV is expected to be at runtime.  This may differ from MYTH_DEFAULT_PREFIX or CMAKE_INSTALL_PREFIX for packagers."
+)
 
 # Location for downloaded tarballs.
 #
diff --git a/mythtv/libs/libmythbase/CMakeLists.txt b/mythtv/libs/libmythbase/CMakeLists.txt
index c51e056175..5bde73d5b9 100644
--- a/mythtv/libs/libmythbase/CMakeLists.txt
+++ b/mythtv/libs/libmythbase/CMakeLists.txt
@@ -261,11 +261,15 @@ target_compile_definitions(mythbase PRIVATE MBASE_API)
 set_source_files_properties(mythcdrom.cpp PROPERTIES COMPILE_DEFINITIONS
                                                      HAVE_LIBUDFREAD)
 
+if(NOT MYTH_RUN_PREFIX)
+  message(VERBOSE "Setting MYTH_RUN_PREFIX to ${CMAKE_INSTALL_PREFIX}")
+  set(MYTH_RUN_PREFIX ${CMAKE_INSTALL_PREFIX})
+endif()
 set_source_files_properties(
   mythdirs.cpp
   PROPERTIES
     COMPILE_DEFINITIONS
-    "RUNPREFIX=\"${CMAKE_INSTALL_PREFIX}\";LIBDIRNAME=\"${CMAKE_INSTALL_LIBDIR}\""
+    "RUNPREFIX=\"${MYTH_RUN_PREFIX}\";LIBDIRNAME=\"${CMAKE_INSTALL_LIBDIR}\""
 )
 target_include_directories(
   mythbase
```

**Comment by amessina** [comment]
> How about this? Does this give you the control you need?
> 
> ```diff
> index 1dfac3b601..ebf22af5ee 100644
> --- a/cmake/MythOptions.cmake
> +++ b/cmake/MythOptions.cmake
> @@ -21,6 +21,11 @@ set(MYTH_DEFAULT_LIBS_PREFIX
>      CACHE PATH
>            "Default library install prefix if not specified on the command line."
>  )
> +set(MYTH_RUN_PREFIX
> +    ""
> +    CACHE PATH
> +          "The prefix where MythTV is expected to be at runtime.  This may differ from MYTH_DEFAULT_PREFIX or CMAKE_INSTALL_PREFIX for packagers."
> +)
>  
>  # Location for downloaded tarballs.
>  #
> diff --git a/mythtv/libs/libmythbase/CMakeLists.txt b/mythtv/libs/libmythbase/CMakeLists.txt
> index c51e056175..5bde73d5b9 100644
> --- a/mythtv/libs/libmythbase/CMakeLists.txt
> +++ b/mythtv/libs/libmythbase/CMakeLists.txt
> @@ -261,11 +261,15 @@ target_compile_definitions(mythbase PRIVATE MBASE_API)
>  set_source_files_properties(mythcdrom.cpp PROPERTIES COMPILE_DEFINITIONS
>                                                       HAVE_LIBUDFREAD)
>  
> +if(NOT MYTH_RUN_PREFIX)
> +  message(VERBOSE "Setting MYTH_RUN_PREFIX to ${CMAKE_INSTALL_PREFIX}")
> +  set(MYTH_RUN_PREFIX ${CMAKE_INSTALL_PREFIX})
> +endif()
>  set_source_files_properties(
>    mythdirs.cpp
>    PROPERTIES
>      COMPILE_DEFINITIONS
> -    "RUNPREFIX=\"${CMAKE_INSTALL_PREFIX}\";LIBDIRNAME=\"${CMAKE_INSTALL_LIBDIR}\""
> +    "RUNPREFIX=\"${MYTH_RUN_PREFIX}\";LIBDIRNAME=\"${CMAKE_INSTALL_LIBDIR}\""
>  )
>  target_include_directories(
>    mythbase
> ```

I think that would work, though I'll need to wait until this weekend to try it. The following also references the resulting run prefix so that may be another place to use `MYTH_RUN_PREFIX` when set. https://github.com/MythTV/mythtv/blob/683d1a4e14c226fbe90577c8da888e4eb95f08ae/mythtv/cmake/PrintConfig.cmake#L63

**Comment by linuxdude42** [comment]
Let me know how the testing goes.  I'll leave the issue open for now in case you need further tweaks.

**Comment by amessina** [comment]
> Let me know how the testing goes. I'll leave the issue open for now in case you need further tweaks.

I was able to rebuild with 31ea30ccf7ef529c5ec359b189e176ec9b8ec989 atop fixes/34 and the patch works beautifully.  If this qualifies for backport, I'd appreciate it.

Thank you @linuxdude42.

**Comment by amessina** [comment]
Reopening for additional information from other packagers and testing.  See comments on 257c12fd550d06d546064d2ca5fd780ceca65088.

Also note that my references to `DESTDIR` are the equivalent of MythTV's `INSTALL_ROOT` in the configure/make/make install method: essentially they both allow the full installation to be placed somewhere (where package builds will then wrap it up).

**Comment by bennettpeter** [comment]
The location for the full installation is CMAKE_INSTALL_PREFIX. For example this is what I use:

```
cmake --preset qt5 \
-DMYTH_BUILD_PLUGINS=ON \
-DCMAKE_BUILD_TYPE=Debug \
-DCMAKE_INSTALL_PREFIX=/home/peter/proj/build/noble-amd64/mythtv/cmake/usr \
|& tee ../cmake1.log

cmake --build build-qt5 |& tee ../cmake2.log

```
This puts the entire build into /home/peter/proj/build/noble-amd64/mythtv/cmake/usr and I should be able to either run it from there or package that directory. and install it into /usr on another computer and run from there on the installed computer. Do I need a MYTH_RUN_PREFIX as well?

**Comment by amessina** [comment]
> The location for the full installation is CMAKE_INSTALL_PREFIX. For example this is what I use:
> 
> ```
> cmake --preset qt5 \
> -DMYTH_BUILD_PLUGINS=ON \
> -DCMAKE_BUILD_TYPE=Debug \
> -DCMAKE_INSTALL_PREFIX=/home/peter/proj/build/noble-amd64/mythtv/cmake/usr \
> |& tee ../cmake1.log
> 
> cmake --build build-qt5 |& tee ../cmake2.log
> ```
> 
> This puts the entire build into /home/peter/proj/build/noble-amd64/mythtv/cmake/usr and I should be able to either run it from there or package that directory. and install it into /usr on another computer and run from there on the installed computer. Do I need a MYTH_RUN_PREFIX as well?

The same thing *kind of* worked for me for Fedora 41, but inspired me to report this issue to improve support for packaging/buildsystem builds.  Once the `MYTH_RUN_PREFIX` option was added, everything seemed to work ok but I reopened the ticket as there are still some nuances as noted by others in the comments on commit 257c12fd550d06d546064d2ca5fd780ceca65088.

To be package compatible, `CMAKE_INSTALL_PREFIX` should be able to default to `/usr` and the installation should go into `INSTALL_ROOT` or more commonly `DESTDIR` (just like the old Makefile style did).

**Comment by rcrdnalor** [comment]
Facts:
MythTV uses the configured `RUNPREFIX` in
libs/libmythbase/mythdirs.cpp (initializeMythDirs()) to set-up its core-context.
The `RUNPREFIX` defaults to `prefix` (`/usr/local`) on make builds
and on cmake builds it defaults to `CMAKE_INSTALL_PREFIX` (`/usr/local`)
Both, cmake and make initialize the `RUNPREFIX` correctly.

The Python Bindings use something similar to detect the installation paths
in bindings/python/MythTV/_versions.py (`MYTHTV_INSTALL_PREFIX`)
which defaults to `prefix` (make) or `CMAKE_INSTALL_PREFIX` (cmake)
In case of `CMAKE_INSTALL_PREFIX` is different to the `MYTH_RUN_PREFIX`
the `MYTHTV_INSTALL_PREFIX` is set wrong, see 
https://github.com/MythTV/mythtv/blob/d6faefe707ec3db14969e1f3c099a61b56b663a5/mythtv/bindings/python/CMakeLists.txt#L86
This should be set to 
`set(MYTHTV_INSTALL_PREFIX "${MYTH_RUN_PREFIX}")

Both are configured either by `configure` step for `make` or by the `cmake --preset` step.

In other words, answering the question from Peter:
If you build for Debian Packages: yes, the `MYTH_RUN_PREFIX` needs to be set.
In your case to `/usr` according MythTV/packaging/deb-light.
I guess this also applies to the rpm builds.
If  you build for testing (using `test.sh` in deb-light),
the `MYTH_RUN_PREFIX` needs to be unset, but the `CMAKE_INSTALL_PREFIX`
needs to be set to the accordingly (needed by `RUNPREFIX`).

**Comment by amessina** [comment]
@rcrdnalor thank you. `set(MYTHTV_INSTALL_PREFIX "${MYTH_RUN_PREFIX}")` does update the `INSTALL_PREFIX` of the `_versions.py` file:

```
~# cat /usr/lib/python3.13/site-packages/MythTV/_versions.py 
# -*- coding: utf-8 -*-

"""
Contains static and global variables for MythTV Python Bindings
related to version information.
The file MythTV/_versions.py is usually generated by ./configure.
"""

OWN_VERSION = (34,0,-1,0)
SCHEMA_VERSION = 1379
NVSCHEMA_VERSION = 1007
MUSICSCHEMA_VERSION = 1025
PROTO_VERSION = '91'
PROTO_TOKEN = 'BuzzOff'
BACKEND_SEP = '[]:[]'
INSTALL_PREFIX = '/usr'
```
